classes/file_search.php

```diff
@@ -0,0 +1,423 @@
+    *
+    * @var int
+    */
+    const CSV_FILENAME  = 5;
+
+    /**
+    * 7th column of csv output - the mimetype column of the mdl_files table.
+    *
+    * @var int
+    */
+    const CSV_MIMETYPE  = 6;
+
+    /**
+    * 8th column of csv output - the strategy used to search the file.
+    *
+    * @var int
+    */
+    const CSV_STRATEGY  = 7;
+
+    /**
+    * 9th column of csv output - some internal information, depending on the strategy.
+    *
+    * For example, for zip strategy, this will be path used inside the zip for the subfile being searched.
+    *
+    * @var int
+    */
+    const CSV_INTERNAL  = 8;
+
+    /**
+    * 10th column of csv output - the offset of the match within the file.
+    *
+    * @var int
+    */
+    const CSV_OFFSET    = 9;
+
+    /**
+    * 11th column of csv output - the text that was matched.
+    *
+    * @var int
+    */
+    const CSV_MATCH     = 10;
+
+    /**
+    * 12th column of csv file - the replacement text.
+    *
+    * @var int
+    */
+    const CSV_REPLACE   = 11;
+
+    /**
+     * Searches the DB using a persistent record.
+     *
+     * @param \tool_advancedreplace\files $record
+     * @param string $output path
+     * @return void
+     */
+    public static function files(\tool_advancedreplace\files $record, string $output = ''): void {
+        global $DB;
+        $id = $record->get('id');
+        $pattern = trim($record->get('pattern'));
+        $components = trim($record->get('components'));
+        $skipcomponents = trim($record->get('skipcomponents'));
+        $mimetypes = trim($record->get('mimetypes'));
+        $skipmimetypes = trim($record->get('skipmimetypes'));
+        $filenames = trim($record->get('filenames'));
+        $skipfilenames = trim($record->get('skipfilenames'));
+        $skipareas = trim($record->get('skipareas'));
+        $origin = trim($record->get('origin'));
+        $filename = \tool_advancedreplace\files::get_filename($record->to_record());
+        // Create temp output directory.
+        if (!$output) {
+            $dir = make_request_directory();
+            $output = $dir . '/' . $filename;
+        }
+
+        // Start output.
+        $stream = fopen($output, 'w');
+
+        $columnheaders = [
+            'contextid', 'component', 'filearea', 'itemid', 'filepath', 'filename',
+            'mimetype', 'strategy', 'internal', 'offset', 'match', 'replace',
+        ];
+        fputcsv($stream, $columnheaders);
+
+        $pattern = '/' . $pattern . '/i';
+
+        [$whereclause, $params] =
+        self::make_where_clause($components, $skipcomponents, $skipareas,
+        $mimetypes, $skipmimetypes, $filenames, $skipfilenames);
+        $record->set('timestart', time());
+        $updatetime = time();
+        $updatepercent = 0;
+        $matchcount = 0;
+        $filecount = 0;
+        $totalfiles = 1000; // Todo: get an estimate with a COUNT(*) query.
```

Comment:
This estimate makes the in progress status a bit whack - I was sitting at 51826.3% at one point.